File: 143. Reorder List.cpp
/*
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
//Solution 1: Pass OJ, but not in-place. Actually, wrong answer.
class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return;
        int length = 0; ListNode *temp0 = head;
        while (temp0 != nullptr) {
            temp0 = temp0->next;
            length++;
        }
        ListNode **temp1 = new ListNode *[length];
        temp0 = head;
        for (int i = 0; i < length; i++) {
            temp1[i] = temp0;
            temp0 = temp0->next;
        }
        if (length % 2 == 0) {
            int i = 0;
            for (; i < length/2 - 1; i++) {
                temp1[i]->next = temp1[length - i - 1];
                temp1[length - i - 1]->next = temp1[i + 1];
            }
            temp1[i]->next = temp1[length - i - 1];
            temp1[length - i - 1]->next = nullptr;
        }
        else {
            int i = 0;
            for (; i < length/2; i++) {
                temp1[i]->next = temp1[length - i - 1];
                temp1[length - i - 1]->next = temp1[i + 1];
            }
            temp1[i]->next = nullptr;
        }
    }
};

// Solution 2: reverse second part and link first and second part.
class Solution {
public:
    void reorderList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return;
        //calculate length
        int length = 0; ListNode *first = head, *second;
        while (first != nullptr) {
            first = first->next;
            length++;
        }
        //make partition
        first = head;
        for (int i = 0; i < length / 2 - 1; i++)
            first = first->next;
        if (length % 2 == 0) {
            second = first->next;
            first->next = nullptr; first = head;
        } 
        else {
            second = first->next->next;
            first->next->next = nullptr; first = head;
        }
            
        //reverse second part
        ListNode *prev = nullptr; ListNode *curr = second, *nxt = second->next;
        while (curr != nullptr) {
            curr->next = prev;
            prev = curr;
            curr = nxt;
            if (nxt != nullptr)
                nxt = nxt->next;
        }
        second = prev;
        //link first and second part
        ListNode *first_nxt, *second_nxt;
        for (int i = 0; i < length / 2; i++) {
            first_nxt = first->next; second_nxt = second->next;
            first->next = second;
            second->next = first_nxt;
            first = first_nxt;
            second = second_nxt;
        }
        if (first_nxt != nullptr)
            first_nxt->next = nullptr;
    }
};

File: Linked List/143. Reorder List.py
# Given a singly linked list L: L0→L1→…→Ln-1→Ln,
# reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…
#
# You may not modify the values in the list's nodes, only nodes itself may be changed.
#
# Example 1:
#
# Given 1->2->3->4, reorder it to 1->4->2->3.
# Example 2:
#
# Given 1->2->3->4->5, reorder it to 1->5->2->4->3.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        # find the middle point of the linked list
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        # reverse the second part
        prev = slow
        start = prev.next
        then = start.next
        while then:
            start.next = then.next
            then.next = prev.next
            prev.next = then
            then = start.next
        # reorder one by one
        l1, l2 = head, slow.next
        while l1 != slow:
            slow.next = l2.next
            l1Nxt = l1.next
            l2Nxt = l2.next
            l2.next = l1.next
            l1.next = l2
            l1 = l1Nxt
            l2 = l2Nxt